tools/ci-build/runtime-versioner/src/audit.rs

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::{
+    command::CommandExt,
+    index::CratesIndex,
+    repo::Repo,
+    tag::{previous_release_tag, release_tags},
+    util::utf8_path_buf,
+    Audit,
+};
+use anyhow::{anyhow, bail, Context, Result};
+use camino::{Utf8Path, Utf8PathBuf};
+use smithy_rs_tool_common::release_tag::ReleaseTag;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fs,
+    process::Command,
+};
+
+pub fn audit(args: Audit) -> Result<()> {
+    let repo = Repo::new(args.smithy_rs_path.as_deref())?;
+    if !args.no_fetch {
+        // Make sure we have the latest release tags
+        fetch_smithy_rs_tags(&repo)?;
+    }
+
+    let release_tags = release_tags(&repo)?;
+    let previous_release_tag =
+        previous_release_tag(&repo, &release_tags, args.previous_release_tag.as_deref())?;
+    if release_tags.first() != Some(&previous_release_tag) {
+        tracing::warn!("there are newer releases since '{previous_release_tag}'");
+    }
+
+    let next_crates = discover_runtime_crates(&repo.root).context("next")?;
+    let previous_crates = resolve_previous_crates(&repo, previous_release_tag.as_str())?;
+
+    let crates = augment_runtime_crates(previous_crates, next_crates, args.fake_crates_io_index)?;
+    let mut errors = Vec::new();
+    for rt_crate in crates {
+        if let Err(err) = audit_crate(&repo, &previous_release_tag, rt_crate) {
+            errors.push(err);
+        }
+    }
+    if errors.is_empty() {
+        println!("SUCCESS");
+        Ok(())
+    } else {
+        for error in errors {
+            eprintln!("{error}");
+        }
+        Err(anyhow!("there are audit failures in the runtime crates"))
+    }
+}
+
+fn audit_crate(repo: &Repo, release_tag: &ReleaseTag, rt_crate: RuntimeCrate) -> Result<()> {
+    if rt_crate.changed_since_release(repo, release_tag)? {
+        // If this version has never been published before, then we're good.
+        // (This tool doesn't check semver compatibility.)
+        if !rt_crate.next_version_is_published() {
+            if let Some(previous_version) = rt_crate.previous_release_version {
+                tracing::info!(
+                    "'{}' changed and was version bumped from {previous_version} to {}",
+                    rt_crate.name,
+                    rt_crate.next_release_version,
+                );
+            } else {
+                tracing::info!(
+                    "'{}' is a new crate (or wasn't independently versioned before) and will publish at {}",
+                    rt_crate.name,
+                    rt_crate.next_release_version,
+                );
+            }
+            Ok(())
```

Comment:
I wonder if it would be practical to invoke cargo semverchecks at this point...

Reply:
Given its current inability to work across crates due to rustdoc limitations, it doesn't seem particularly useful.